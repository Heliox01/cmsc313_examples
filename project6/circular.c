#include "circular.h"

/*
 * This is where you should write all your code (except any additional
 * tests you choose to include).
 */


/*
 * First define what the real _cbuf struct should look like. Remember,
 * this is opaque to the users of your library so you can do whatever
 * you like in here. The _cbuf struct is typedefd in the header to the
 * cbuf type so you're free to use either type mostly
 * interchangably. The one exception is that sizeof calls must use the
 * _cbuf type.
 */

struct _cbuf {
  // whatever you want can go in here
};

/*
 * Now define your implementations of all of the functions given in
 * the header. The test programs will not link successfully until each
 * function they use is defined here.
 */
