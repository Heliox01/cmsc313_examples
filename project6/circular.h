#ifndef CIRCULAR_H
#define CIRCULAR_H

/* provides size_t */
#include <stddef.h>

/*
 * This is an opaque data structure. You may define the _cbuf struct
 * inside your C file and include whatever fields you like. The user
 * of your library should never know or care about the internals of
 * your cbuf struct.
 */
struct _cbuf;
typedef struct _cbuf cbuf;

/*
 * This is the type that users will store inside your data
 * structure. Its definition is exposed to users because they will
 * need to be able to access its fields.
 *
 * Your implementation should not need to modify the provided
 * definition of this datastructure.
 */
typedef struct {
  double rate;
  unsigned int time;
} quote;


/*
 * These are the "public" functions used for interacting with your
 * datastructure. You are free to define more functions to be used as
 * internal implementation details but you should define them only in
 * your C file so that the users of your datastructure will not need
 * to know or care about them. If you wish, you can create a seperate
 * "circular_internal.h" header that has definitions of these internal
 * functions so that your tests can access them.
 */


/*
 * Initialize a circular buffer and return a pointer that can be used
 * to specify that circular buffer in future function calls.
 */
cbuf* cbuf_alloc();

/*
 * Deallocate all dynamically allocated memory associated with the
 * circular buffer referenced by cb_ptr. The programmer using your ADT
 * promises to never use the value in cb_ptr again.
 */
void cbuf_free(cbuf *cb_ptr);

/*
 * Add a new currency exchange quote to the specified circular
 * buffer. The time is in seconds since midnight. This function should
 * remove any quotes in the circular buffer that are more than 5
 * minutes older than this quote. You should assume that the
 * timestamps given to cbuf_update() are in non-decreasing
 * order. During a cbuf_update(), the memory for the circular buffer
 * may have to grow or shrink according to the rules given in the
 * project description.
 */
void cbuf_update(cbuf *cb_ptr, unsigned int time, double rate);

/*
 * Return the average exchange rate of the quotes currently stored in
 * the buffer.
 */
double cbuf_average(cbuf *cb_ptr);

/*
 * Return a pointer to the earliest quote currently stored in the
 * buffer. (Earliest = smallest timestamp) The programmer using your
 * ADT promises to just "look at" the quote and not alter it in any
 * way.
 */
const quote *cbuf_start(cbuf *cb_ptr);

/*
 * Return a pointer to the latest quote currently stored in the
 * buffer. (Latest = largest timestamp) The programmer using your ADT
 * promises to just "look at" the quote and not alter it in any way.
 */
const quote *cbuf_end(cbuf *cb_ptr);

/*
 * Return the number of elements currently stored in the circular
 * buffer.
 */
size_t cbuf_size(cbuf *cb_ptr);

/*
 * Return the current capacity of the circular buffer (will be greater
 * than or equal to the number of elements in the buffer). When
 * cbuf_size(x) == cbuf_capacity(x) then your circular buffer may need
 * to grow when the next element is inserted.
 */
size_t cbuf_capacity(cbuf *cb_ptr);

#endif
