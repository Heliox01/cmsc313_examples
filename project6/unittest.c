#include "unittest.h"

#include <stdio.h>

void check(int test, const char* message) {
  printf("Test `%s': %s\n", message, test ? "PASS" : "FAIL");
}
