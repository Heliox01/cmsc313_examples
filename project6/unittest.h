#ifndef UNITTEST_H
#define UNITTEST_H

/*
 * Print the message along with a pass/fail
 */
void check(int test, const char* message);

#endif
