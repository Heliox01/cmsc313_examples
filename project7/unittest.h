#ifndef UNITTEST_H
#define UNITTEST_H

/*
 * Print the message along with a pass/fail
 */
void check(int test, const char* message);

/*
 * Print a message summarizing overall pass/fail. Return 0 on total
 * success.
 */
int check_summary();

#endif
