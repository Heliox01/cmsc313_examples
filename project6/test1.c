/*

 * Test out circular buffer for foreign exchange quotes.
*/

#include <stdio.h>
#include <stdlib.h>
#include "circular.h"
#include "unittest.h"

int main(int argc, char** argv) {
  cbuf *cb1 = cbuf_alloc();

  // make sure the cbuf grows and shrinks appropriately
  int capacity = cbuf_capacity(cb1);
  check(capacity > 0, "Initial capacity > 0");
  check(cbuf_size(cb1) == 0, "Initial size == 0");

  for(int i = 0; i < capacity; i++) {
    cbuf_update(cb1, 60, 1.291);
  }

  check(cbuf_size(cb1) == cbuf_capacity(cb1), "Size is allowed to grow to capacity");

  cbuf_update(cb1, 60, 1.291);

  // inserting one more than what the original structure could contain
  int new_capacity = cbuf_capacity(cb1);
  check(new_capacity > capacity, "Capacity grows when necessary");

  // inserting an update that's > 5 minutes older than all of the data
  // in the structure
  cbuf_update(cb1, 60 + 6 * 60, 1.291);
  check(cbuf_capacity(cb1) < new_capacity, "Capacity shrinks when able");
  check(cbuf_size(cb1) == 1, "Aged out records are removed correctly");

  cbuf_free(cb1);

  return 0;
}
