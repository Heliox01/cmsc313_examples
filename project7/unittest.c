#include "unittest.h"

#include <stdio.h>

static int tests = 0;
static int passes = 0;

void check(int test, const char* message) {
  tests++;
  if(test) passes++;

  printf("[Test %d] `%s': %s\n", tests, message, test ? "PASS" : "FAIL");
}

int check_summary() {
  if(tests == passes) {
    printf("All Tests Pass!\n");
    return 0;
  } else {
    printf("%d out of %d tests did not pass.\n", tests-passes, tests);
    return 1;
  }
}
