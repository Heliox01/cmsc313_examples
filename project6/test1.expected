Test `Initial capacity > 0': PASS
Test `Initial size == 0': PASS
Test `Size is allowed to grow to capacity': PASS
Test `Capacity grows when necessary': PASS
Test `Capacity shrinks when able': PASS
Test `Aged out records are removed correctly': PASS
